### PR TITLE
Add Merkle Tree–Based Chunk-Level Hashing for Dataset Verification this establishes dataset verification integrity .

### LINKED ISSUE #9: Dataset verification limited to single SHA256 hash (no chunk-level integrity support)
### Problem Statement

1. The existing pipeline only generated a single SHA256 checksum for:

- Raw Wikipedia dump

2. Processed cleaned dataset

# Current approach has limitations:

- Requires re-hashing the entire dataset to verify integrity

- Not scalable for large datasets

- No support for partial verification

- No cryptographic structure for subset validation

- For large training corpora, this becomes inefficient and impractical.

### Proposed Solution

1. Implemented a Merkle Tree–based hashing system that:

2. Splits files into fixed-size chunks (default: 1MB)

3. Computes SHA256 hash for each chunk (leaf nodes)

4. Builds a Merkle Tree using raw-byte concatenation

5. Produces a Merkle Root representing the entire dataset

6. Stores the Merkle Root in the dataset manifest

1. This allows cryptographic verification of specific chunks without re-hashing the entire file.

### Implementation Details

- Added `compute_merkle_root()` function

- Reads file in chunks

- Hashes each chunk using SHA256

- Builds Merkle Tree bottom-up

- Returns final Merkle Root (hex string)

### Updated `generate_manifest()` to include:

- `raw_merkle_root`

- `processed_merkle_root`

`chunk_size_bytes`

### Manifest Example (New Fields)
``` json
{
  "raw_sha256": "...",
  "processed_sha256": "...",
  "raw_merkle_root": "...",
  "processed_merkle_root": "...",
  "chunk_size_bytes": 1048576
}
```
### Benefits

• Enables chunk-level verification
• Scales efficiently for large datasets
• Maintains deterministic hashing
• Improves reproducibility
• Aligns with blockchain-style verification models
• Backward compatible

### Testing

1. Added and validated tests for:

2. Merkle root determinism

4. Content modification detection

5. Single-chunk compatibility with SHA256

6. Empty file handling

7. Manifest field presence

All tests pass:

#### Impact

- This change upgrades the pipeline from simple file hashing to structured cryptographic verification.

- It significantly improves scalability and verification robustness for large training datasets while preserving existing functionality.

## Checklist

- [x] My code follows the project's code style and conventions
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings or errors
- [x] I have joined the [Discord server](https://discord.gg/hjUhu33uAn) and I will share a link to this PR with the project maintainers there
- [x] I have read the [Contributing Guidelines](./CONTRIBUTING.md)

## ⚠️ AI Notice - Important!

 We encourage contributors to use AI tools responsibly when creating Pull Requests. While AI can be a valuable aid, it is essential to ensure that your contributions meet the task requirements, build successfully, include relevant tests, and pass all linters. Submissions that do not meet these standards may be closed without warning to maintain the quality and integrity of the project. Please take the time to understand the changes you are proposing and their impact.
 
 
@Archit381 Please check this out

### Steps to Reproduce

N/A

### Logs and Screenshots

N/A

### Environment Details

N/A

### Impact

Medium - Feature works but has issues

### Code of Conduct

- [x] I have joined the [Discord server](https://discord.gg/hjUhu33uAn) and will post updates there
- [x] I have searched existing issues to avoid duplicates

### PR DESCRIPTION
**Pull request by aniket866**
Closes #9 
### Summary

**This PR introduces Merkle Tree–based chunk-level hashing to enhance dataset integrity verification.**

**Previously, the pipeline computed a single SHA256 hash for the entire raw and processed dataset. While this ensured file-level integrity, it required re-hashing the full dataset (potentially multi-gigabyte) to verify any portion of the data.**

- With this update, the system now supports chunk-level cryptographic verification using a Merkle Tree construction.

### Problem Statement

1. The existing pipeline only generated a single SHA256 checksum for:

- Raw Wikipedia dump

2. Processed cleaned dataset

# Current approach has limitations:

- Requires re-hashing the entire dataset to verify integrity

- Not scalable for large datasets

- No support for partial verification

- No cryptographic structure for subset validation

- For large training corpora, this becomes inefficient and impractical.

### Proposed Solution

1. Implemented a Merkle Tree–based hashing system that:

2. Splits files into fixed-size chunks (default: 1MB)

3. Computes SHA256 hash for each chunk (leaf nodes)

4. Builds a Merkle Tree using raw-byte concatenation

5. Produces a Merkle Root representing the entire dataset

6. Stores the Merkle Root in the dataset manifest

1. This allows cryptographic verification of specific chunks without re-hashing the entire file.

### Implementation Details

- Added `compute_merkle_root()` function

- Reads file in chunks

- Hashes each chunk using SHA256

- Builds Merkle Tree bottom-up

- Returns final Merkle Root (hex string)

### Updated `generate_manifest()` to include:

- `raw_merkle_root`

- `processed_merkle_root`

`chunk_size_bytes`

### Manifest Example (New Fields)
``` json
{
  "raw_sha256": "...",
  "processed_sha256": "...",
  "raw_merkle_root": "...",
  "processed_merkle_root": "...",
  "chunk_size_bytes": 1048576
}
```
### Benefits

• Enables chunk-level verification
• Scales efficiently for large datasets
• Maintains deterministic hashing
• Improves reproducibility
• Aligns with blockchain-style verification models
• Backward compatible

### Testing

1. Added and validated tests for:

2. Merkle root determinism

4. Content modification detection

5. Single-chunk compatibility with SHA256

6. Empty file handling

7. Manifest field presence

All tests pass:

### 16 passed in 1.88s
image

#### Impact

- This change upgrades the pipeline from simple file hashing to structured cryptographic verification.

- It significantly improves scalability and verification robustness for large training datasets while preserving existing functionality.

## Checklist

- [x] My code follows the project's code style and conventions
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings or errors
- [x] I have joined the [Discord server](https://discord.gg/hjUhu33uAn) and I will share a link to this PR with the project maintainers there
- [x] I have read the [Contributing Guidelines](./CONTRIBUTING.md)

## ⚠️ AI Notice - Important!

 We encourage contributors to use AI tools responsibly when creating Pull Requests. While AI can be a valuable aid, it is essential to ensure that your contributions meet the task requirements, build successfully, include relevant tests, and pass all linters. Submissions that do not meet these standards may be closed without warning to maintain the quality and integrity of the project. Please take the time to understand the changes you are proposing and their impact.
 
 
@Archit381 Please check this out

## Summary by CodeRabbit

* **New Features**
  * Added Merkle-root based cryptographic checksums for file verification.
  * Manifest now includes Merkle root hashes and chunk-size metadata for raw and processed files.
  * Hashing enhanced to accept additional input formats for verification.

* **Tests**
  * Added tests for Merkle determinism, content-change detection, single-chunk and empty-file edge cases.

### CODE REVIEW

File: tests/test_merkle.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+def test_merkle_root_empty_file(tmp_path):
+    file = tmp_path / "empty.txt"
+    file.write_text("")
+
+    root = utils.compute_merkle_root(file)
+
+    expected = hashlib.sha256(b"").hexdigest()
+
+    assert root == expected
```
[the remark this answers is not in the document]
Comment (aniket866) [reply tests/test_merkle.py]: That can be added later when test refactoring is to be carried out , I will do it later

File: openverifiablellm/utils.py [hunk 2]
```diff
@@ -74,6 +121,13 @@ def generate_manifest(raw_path, processed_path):
         "dump_date": extract_dump_date(raw_path.name),
         "raw_sha256": compute_sha256(str(raw_path)),
         "processed_sha256": compute_sha256(str(processed_path)),
+
+        # ---------------- ADDED FIELDS ----------------
+        "raw_merkle_root": compute_merkle_root(raw_path),
+        "processed_merkle_root": compute_merkle_root(processed_path),
+        "chunk_size_bytes": 1024 * 1024,
+        # ---------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (aniket866) [reply openverifiablellm/utils.py]: solved

**Comment by aniket866** [comment]
Hii @Archit381 I have made the changes you suggested , Please have a look

**Comment by Archit381** [comment]
@aniket866 data doesnt seem the best if someone has to pass a `file_path` for loading. Maybe add another param, `file_path` as an Optional
